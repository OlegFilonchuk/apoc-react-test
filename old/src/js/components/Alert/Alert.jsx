import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { DEFAULT_BACKDROP_ZINDEX } from '../Backdrop/Backdrop';

export const ALERT_TYPES = {
  WARNING: 'warning',
  ERROR: 'error',
  SUCCESS: 'success'
};

const ALERT_TYPE_TO_CLASS = {
  [ALERT_TYPES.WARNING]: 'sc-alert-warning',
  [ALERT_TYPES.ERROR]: 'sc-alert-error',
  [ALERT_TYPES.SUCCESS]: 'sc-alert-success'
};

export default class Alert extends Component {
  constructor(props) {
    super(props);

    this.close = this.close.bind(this);
  }

  componentWillMount() {
    this.setStateFromProps();
  }

  componentDidMount() {
    if (this.props.closeTimeout) {
      this.intervalId = setTimeout(this.close, this.props.closeTimeout * 1000);
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.isOpen !== this.state.isOpen) {
      this.setStateFromProps();
    }
  }

  componentWillUnmount() {
    window.clearInterval(this.intervalId);
  }

  setStateFromProps() {
    this.setState((previousState, props) => ({ isOpen: props.isOpen }));
  }

  close() {
    this.setState({ isOpen: false }, this.props.onClose);
  }

  render() {
    const alertClassName = `sc-alert sc-alert-closable sc-animate alert ${ALERT_TYPE_TO_CLASS[this.props.type]} ${
      this.props.className
    }`;

    const closeButton = this.props.isClosable ? (
      <button className="sc-close" aria-label="Close" onClick={this.close}>
        <span aria-hidden="true" className="sc-icon-xmark" />
      </button>
    ) : null;

    const { zIndex } = this.props;

    return this.state.isOpen ? (
      <div className={alertClassName} role="alert" style={{ zIndex }}>
        <div className="alert__label">{this.props.children}</div>

        {closeButton}
      </div>
    ) : null;
  }
}

Alert.propTypes = {
  type: PropTypes.oneOf([ALERT_TYPES.SUCCESS, ALERT_TYPES.WARNING, ALERT_TYPES.ERROR]),

  /**
   * Whether alert should be opened or closed by default
   */
  isOpen: PropTypes.bool,
  /* eslint react/no-unused-prop-types: 0 */

  /**
   * Should alert have close button
   */
  isClosable: PropTypes.bool,

  /**
   * Callback to be called after Alert will be closed
   */
  onClose: PropTypes.func,

  /**
   * Can be either simple text or link
   */
  children: PropTypes.node.isRequired,

  /**
   * Time, expressed in seconds, after which Alert will close, set `0` to prevent autoclosing
   */
  closeTimeout: PropTypes.number,

  /**
   * Any custom class names
   */
  className: PropTypes.string,

  /**
   * Allow to directly override `z-index` for Alert
   */
  zIndex: PropTypes.number
};

Alert.defaultProps = {
  type: ALERT_TYPES.SUCCESS,
  isOpen: true,
  isClosable: true,
  onClose: () => true,
  closeTimeout: 3,
  className: '',
  zIndex: DEFAULT_BACKDROP_ZINDEX + 1
};
