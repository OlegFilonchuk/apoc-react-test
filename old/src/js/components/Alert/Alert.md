## Basic Alert

Simple usage of the **Alert** component:

    (
      <div>
        <Alert closeTimeout={0}>You have an alert!</Alert>
      </div>
    )

## Types & Colors

Alert component can support different types / colors:

    (
      <div>
        <Alert closeTimeout={0} type="success">
          <span className="sc-icon-check" />
          You have a success alert!
        </Alert>

        <Alert closeTimeout={0} type="warning">You have a warning alert!</Alert>

        <Alert closeTimeout={0} type="error">
          <span className="sc-icon-triangle-warning" />You have an error alert!
        </Alert>
      </div>
    )

## Closing timeout

Alert component will close by default in 3 seconds.
Pass `0` if you don't want to close it automatically.

    const initialState = {
      isVisible: false
    };

    const onClickFn = () => {
      setState({ isVisible: true });
    };

    (
      <div>
        <Button onClick={onClickFn}>Click to show auto-closing alert</Button>
        {state.isVisible ? <Alert>I'll be closed in 3 seconds.</Alert> : null}
      </div>
    )

## onClose callback

An `onClose` callback will be called after component will close, programatically or automatically.

    const initialState = {
      isVisible: false
    };

    const onClickFn = () => {
      setState({ isVisible: true });
    };

    const onCloseFn = () => { console.log('Alert closed') };

    (
      <div>
        <Button onClick={onClickFn}>Click to show auto-closing alert</Button>
        {
          state.isVisible ?
            <Alert onClose={onCloseFn} closeTimeout={10} type="success">Alert will close after 10 seconds</Alert> : null
        }
        <Alert onClose={onCloseFn} closeTimeout={0} type="success">Close me with `x` on right!</Alert>
      </div>
    )

## Non-closable alert

    (
      <div>
        <Alert closeTimeout={0} isClosable={false}>You can't close me.</Alert>
      </div>
    )

## Alert closed by default

    const initialState = {
      isOpen: false
    };

    const onClickFn = () => {
      setState({ isOpen: !state.isOpen });
    };

    (
      <div>
        <Button onClick={onClickFn}>Show/hide `Alert`</Button>
        <Alert isOpen={state.isOpen} closeTimeout={0} isClosable={false}>You can close meÂ programatically.</Alert>
      </div>
    )

## Normal `Alert` with `withBackdrop` HoC
    const withBackdrop = require('../withBackdrop/withBackdrop').default;
    const AlertWithBackdrop = withBackdrop(Alert, 'isOpen');

    const initialState = {
      isOpen: false
    };

    const onClickFn = () => {
      setState({ isOpen: true });
    };

    const closeAlert = () => {
      setState({ isOpen: false });
    };

    const onCloseFn = () => {
      console.log('Alert closed')
    };

    (
      <div>
        <Button onClick={onClickFn}>Click to show auto-closing alert</Button>
        <AlertWithBackdrop
          isOpen={state.isOpen}
          backdrop={{ onClick: closeAlert }}
          onClose={onCloseFn}
          closeTimeout={0}
          type="success"
        >
          AlertWithBackdrop with backdrop
        </AlertWithBackdrop>
      </div>
    )
