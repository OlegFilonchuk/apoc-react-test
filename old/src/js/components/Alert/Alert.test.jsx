import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import Alert from './Alert';

const mountAlert = (props = {}, mountMethod = shallow) => mountMethod(<Alert {...props}>Text</Alert>);

describe('<Alert>', () => {
  it('should render opened <Alert>', () => {
    expect(toJson(mountAlert())).toMatchSnapshot();
  });

  it('should render closed <Alert>', () => {
    expect(toJson(mountAlert({ isOpen: false }, mount))).toMatchSnapshot();
  });

  it('should render non closable <Alert>', () => {
    expect(toJson(mountAlert({ isClosable: false }))).toMatchSnapshot();
  });

  it('should fire onClose event when clicking on close button', () => {
    const onClose = jest.fn();
    const alert = mountAlert({ onClose }, mount);

    alert.find('button.sc-close').simulate('click');

    expect(onClose).toHaveBeenCalled();
  });

  it('should fire onClose event after 3 seconds timeout by default', () => {
    // Enables cheats
    jest.useFakeTimers();

    // Our callback to be called
    const onClose = jest.fn();

    // Render
    mountAlert({ onClose }, mount);

    // At this point in time, the callback should not have been called yet
    expect(onClose).not.toBeCalled();

    // Fast forward
    jest.runAllTimers();

    expect(onClose).toHaveBeenCalled();
  });
});
