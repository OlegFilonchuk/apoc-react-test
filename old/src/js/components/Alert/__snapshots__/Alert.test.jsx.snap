// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Alert> should render closed <Alert> 1`] = `
<Alert
  className=""
  closeTimeout={3}
  isClosable={true}
  isOpen={false}
  onClose={[Function]}
  type="success"
  zIndex={15501}
/>
`;

exports[`<Alert> should render non closable <Alert> 1`] = `
<div
  className="sc-alert sc-alert-closable sc-animate alert sc-alert-success "
  role="alert"
  style={
    Object {
      "zIndex": 15501,
    }
  }
>
  <div
    className="alert__label"
  >
    Text
  </div>
</div>
`;

exports[`<Alert> should render opened <Alert> 1`] = `
<div
  className="sc-alert sc-alert-closable sc-animate alert sc-alert-success "
  role="alert"
  style={
    Object {
      "zIndex": 15501,
    }
  }
>
  <div
    className="alert__label"
  >
    Text
  </div>
  <button
    aria-label="Close"
    className="sc-close"
    onClick={[Function]}
  >
    <span
      aria-hidden="true"
      className="sc-icon-xmark"
    />
  </button>
</div>
`;
